src/integration-test/java/com/commercetools/sync/integration/sdk2/ctpprojectsource/products/ProductReferenceResolverIT.java

```diff
@@ -0,0 +1,432 @@
+    final ProductSyncStatistics syncStatistics =
+        productSync.sync(productDrafts).toCompletableFuture().join();
+
+    // assertion
+    assertThat(syncStatistics).hasValues(1, 1, 0, 0);
+    assertThat(errorCallBackMessages).isEmpty();
+    assertThat(errorCallBackExceptions).isEmpty();
+    assertThat(warningCallBackMessages).isEmpty();
+  }
+
+  @Test
+  void sync_withNewProductWithNoProductTypeKey_ShouldFailCreatingTheProduct() {
+    // preparation
+    final ProductDraft productDraft =
+        createProductDraft(
+            PRODUCT_KEY_1_RESOURCE_PATH,
+            noKeyProductTypeSource.toReference(),
+            oldTaxCategory.toReference(),
+            oldProductState.toReference(),
+            categoryReferencesWithIds,
+            createRandomCategoryOrderHints(categoryReferencesWithIds));
+    CTP_SOURCE_CLIENT.products().create(productDraft).executeBlocking();
+
+    final List<ProductProjection> products =
+        CTP_SOURCE_CLIENT
+            .productProjections()
+            .get()
+            .withStaged(true)
+            .execute()
+            .thenApply(ApiHttpResponse::getBody)
+            .thenApply(ProductProjectionPagedQueryResponse::getResults)
+            .join();
+
+    final List<ProductDraft> productDrafts =
+        ProductTransformUtils.toProductDrafts(CTP_SOURCE_CLIENT, referenceIdToKeyCache, products)
+            .join();
+
+    // test
+    final ProductSyncStatistics syncStatistics =
+        productSync.sync(productDrafts).toCompletableFuture().join();
+
+    // assertion
+    assertThat(syncStatistics).hasValues(1, 0, 0, 1);
+    assertThat(errorCallBackMessages)
+        .containsExactly(
+            format(
+                "Failed to process the ProductDraft with"
+                    + " key:'%s'. Reason: "
+                    + ReferenceResolutionException.class.getCanonicalName()
+                    + ": "
+                    + "Failed to resolve 'product-type' resource identifier on ProductDraft with "
+                    + "key:'%s'. Reason: %s",
+                productDraft.getKey(),
+                productDraft.getKey(),
+                BLANK_KEY_VALUE_ON_RESOURCE_IDENTIFIER));
+    assertThat(errorCallBackExceptions).hasSize(1);
+    assertThat(errorCallBackExceptions.get(0)).isExactlyInstanceOf(CompletionException.class);
+    assertThat(warningCallBackMessages).isEmpty();
+  }
+
+  @Test
+  void sync_withNewProductWithStateReferenceVariantAttribute_ShouldCreateProduct() {
+
+    createProductWithStateReferenceVariantAttribute(oldProductState.getId());
+
+    final List<ProductProjection> products =
+        CTP_SOURCE_CLIENT
+            .productProjections()
+            .get()
+            .withStaged(true)
+            .execute()
+            .thenApply(ApiHttpResponse::getBody)
+            .thenApply(ProductProjectionPagedQueryResponse::getResults)
+            .join();
+
+    final List<ProductDraft> productDrafts = // mapToProductDrafts(products);
```

Comment:
This is commented multiple times. I think you can remove it?

Reply:
https://github.com/commercetools/commercetools-sync-java/pull/937/commits/ce838b0eb9845530ca81fce9fc944139cebf80d0